numpy/core/shape_base.py

```diff
@@ -365,78 +365,91 @@ def stack(arrays, axis=0, out=None):
-                            f_reduce=lambda x, **kwargs: x,
-                            f_kwargs=lambda **kwargs: kwargs,
-                            **kwargs):
-        """
-        Iterate over the nested list, applying:
-        * ``f_map`` (T -> U) to items
-        * ``f_reduce`` (Iterable[U] -> U) to mapped items
-
-        For instance, ``map_reduce([[1, 2], 3, 4])`` is::
-
-            f_reduce([
-              f_reduce([
-                f_map(1),
-                f_map(2)
-              ]),
-              f_map(3),
-              f_map(4)
-            ]])
-
-
-        State can be passed down through the calls with `f_kwargs`,
-        to iterables of mapped items. When kwargs are passed, as in
-        ``map_reduce([[1, 2], 3, 4], **kw)``, this becomes::
-
-            kw1 = f_kwargs(**kw)
-            kw2 = f_kwargs(**kw1)
-            f_reduce([
-              f_reduce([
-                f_map(1), **kw2)
-                f_map(2,  **kw2)
-              ],      **kw1),
-              f_map(3, **kw1),
-              f_map(4, **kw1)
-            ]],     **kw)
-        """
-        def f(x, **kwargs):
-            if not self.recurse_if(x):
-                return f_map(x, **kwargs)
-            else:
-                next_kwargs = f_kwargs(**kwargs)
-                return f_reduce((
-                    f(xi, **next_kwargs)
-                    for xi in x
-                ), **kwargs)
-        return f(x, **kwargs)
-
-    def walk(self, x, index=()):
-        """
-        Iterate over x, yielding (index, value, entering), where
-
-        * ``index``: a tuple of indices up to this point
-        * ``value``: equal to ``x[index[0]][...][index[-1]]``. On the first iteration, is
-                     ``x`` itself
-        * ``entering``: bool. The result of ``recurse_if(value)``
-        """
-        do_recurse = self.recurse_if(x)
-        yield index, x, do_recurse
-
-        if not do_recurse:
-            return
-        for i, xi in enumerate(x):
-            # yield from ...
-            for v in self.walk(xi, index + (i,)):
-                yield v
+    def format_index(index):
+        idx_str = ''.join('[{}]'.format(i) for i in index if i is not None)
+        return 'arrays' + idx_str
+    if type(arrays) is tuple:
+        # not strictly necessary, but saves us from:
+        #  - more than one way to do things - no point treating tuples like
+        #    lists
+        #  - horribly confusing behaviour that results when tuples are
+        #    treated like ndarray
+        raise TypeError(
+            '{} is a tuple. '
+            'Only lists can be used to arrange blocks, and np.block does '
+            'not allow implicit conversion from tuple to ndarray.'.format(
+                format_index(parent_index)
+            )
+        )
+    elif type(arrays) is list and len(arrays) > 0:
+        indexes, arr_ndims = zip(*[_block_check_depths_match(arr, parent_index + [i])
```

Comment:
I suspect this is where things slow - can you just iterate over the full result in the loop below, and use `arr_ndims.append`?

Reply:
I've implemented something that avoids the zip(*...) and pushed it, it seems a little better, though there are still some things that are a little slower than 8a83a5fc, probably just because `_nx.ndim` is called on every item at the bottom of the nesting now, which I think we can't get around without reverting to the previous approach.

The two things that are really significantly slower are trivial use cases and are still very fast relative to other use cases anyway.
```
    before     after       ratio
  [8a83a5fc] [2e948722]
+   11.82μs    19.16μs      1.62  bench_shape_base.Block.time_no_lists(10)
+    9.87μs    15.62μs      1.58  bench_shape_base.Block.time_no_lists(1)
+  460.36μs   588.47μs      1.28  bench_shape_base.Block.time_3d(10)
+   14.19μs    16.98μs      1.20  bench_shape_base.Block.time_block_simple_row_wise(10)
+   54.07μs    64.21μs      1.19  bench_shape_base.Block.time_no_lists(100)
+   80.41μs    88.62μs      1.10  bench_shape_base.Block.time_3d(1)
-  277.39μs   256.95μs      0.93  bench_shape_base.Block.time_nested(100)
-   83.21μs    73.71μs      0.89  bench_shape_base.Block.time_block_simple_row_wise(100)
-   67.95μs    59.97μs      0.88  bench_shape_base.Block.time_block_complicated(10)
-   62.88μs    51.18μs      0.81  bench_shape_base.Block.time_block_complicated(1)
-   25.42μs    20.63μs      0.81  bench_shape_base.Block.time_block_simple_column_wise(1)
-     5.98s      4.77s      0.80  bench_shape_base.Block.time_3d(100)
-   86.93μs    67.27μs      0.77  bench_shape_base.Block.time_nested(10)

Reply:
I.e. in absolute terms there's unlikely to be much slowdown to users' code

Reply:
> probably just because _nx.ndim is called on every item at the bottom of the nesting now

Wasn't this always the case? The difference is, you previously also called it on every intermediate result too.

Does avoiding a generator comprehension speed things up? You could call `_block_check_depths_match` in the `for` loop directly, although I agree what you have is cleaner

Reply:
I was calling `atleast_nd` on every level (including the bottom one), but ndim was only called for depth < max_depth I believe. So perhaps ndim, when called on scalars, is slower than our atleast_nd function (which is just a wrapper for np.array)?